Draft review Nico (#123)
scionassociation/scion-cppki_I-D: draft-dekater-scion-pki.md

- Multi-party governance; and
- Support for policy versioning and updates.

Ideally, the trust architecture allows parties that mutually trust each other to form their own trust domain, and to freely decide whether to trust other trust domains.

> To fulfill the above requirements, which in fact apply well to inter-domain networking, SCION introduces the concept of **Isolation Domains**. An Isolation Domain (ISD) is a building block to support for heterogeneous trust while achieving high availability and scalability in {{I-D.dekater-scion-controlplane}}. It consists of a logical grouping of SCION ASes that share a uniform trust environment (i.e. a common jurisdiction).

Reviewer: "to support heterogeneous trust while.."

Author: ```suggestion
To fulfill the above requirements, which in fact apply well to inter-domain networking, SCION introduces the concept of **Isolation Domains**. An Isolation Domain (ISD) is a building block to support heterogeneous trust while achieving high availability and scalability in {{I-D.dekater-scion-controlplane}}. It consists of a logical grouping of SCION ASes that share a uniform trust environment (i.e. a common jurisdiction).
```

An ISD is governed by one or multiple ASes, known as the **voting ASes**. Furthermore, each ISD has a set of ASes that form the ISD core, known as the **core ASes**. The set of core and voting ASes may be, but do not necessarily have to be the same ASes. Governance is implemented by a policy called the **Trust Root Configuration** (TRC), which is negotiated by the voting ASes, and which defines the locally scoped roots of trust used to validate bindings between names and public keys.

> Authentication in SCION is based on X.509 certificates that bind identifiers to public keys and carry digital signatures that are verified by roots of trust. SCION allows each ISD to define its own set of trust roots, along with the policy governing their use. Such scoping of trust roots within an ISD improves security as compromise of a private key associated with a trust root cannot be used to forge a certificate outside the ISD. An ISD's trust roots and policy are encoded in the TRC, which has a version number, a list of public keys that serves as root of trust for various purposes, and a voting quorums governing the number of signatures required to update TRCs. The TRC serves as a way to bootstrap all authentication within SCION. Additionally, TRC versioning is used to efficiently revoke compromised roots of trust.

Reviewer: "and a voting quorum..."

Author: ```suggestion
Authentication in SCION is based on X.509 certificates that bind identifiers to public keys and carry digital signatures that are verified by roots of trust. SCION allows each ISD to define its own set of trust roots, along with the policy governing their use. Such scoping of trust roots within an ISD improves security as compromise of a private key associated with a trust root cannot be used to forge a certificate outside the ISD. An ISD's trust roots and policy are encoded in the TRC, which has a version number, a list of public keys that serves as root of trust for various purposes, and a voting quorum governing the number of signatures required to update TRCs. The TRC serves as a way to bootstrap all authentication within SCION. Additionally, TRC versioning is used to efficiently revoke compromised roots of trust.
```

~~~~
{: #figure-1 title="Chain of trust within an ISD"}

All certificates used in the Control Plane PKI are in X.509 v3 format {{RFC5280}} and additionally the TRC contains self-signed certificates instead of plain public keys. Self-signed certificates have the following advantages over plain public keys: (1) They make the binding between name and public key explicit; and (2) the binding is signed to prove possession of the corresponding private key. The public keys of Voting AS certificates must therefore be explicitly verified during the Signing Ceremony ([](#initial-ceremony)).

SCION ASes sign and verify control plane messages. Certain ASes have additional roles:

> - **Core ASes**: they are a distinct set of ASes in the SCION Control Plane. For each ISD, the core ASes are listed in the TRC and each core AS has links to the other core ASes (in the same or in different ISDs).

Reviewer: Would capitalise 'They'

Author: ```suggestion
- **Core ASes**: They are a distinct set of ASes in the SCION Control Plane. For each ISD, the core ASes are listed in the TRC and each core AS has links to the other core ASes (in the same or in different ISDs).
```

- **Certification authorities (CAs)**: CAs are responsible for issuing AS certificates to other ASes and/or themselves.

> - **Voting ASes**: they may sign TRC updates. The process of appending a signature to a new TRC is called "casting a vote", and the designated ASes that hold the private keys to sign a TRC update are "voting ASes".

Reviewer: Would capitalise 'They'

Author: ```suggestion
- **Voting ASes**: They may sign TRC updates. The process of appending a signature to a new TRC is called "casting a vote", and the designated ASes that hold the private keys to sign a TRC update are "voting ASes".
```

In SCION, using the `keyIdentifier` attribute is the preferred way to specify the `authorityKeyIdentifier` extension.

SCION implementations MAY also support the use of the `authorityCertIssuer` and `authorityCertSerialNumber` attributes. However, if these attributes are set and support for them is missing, implementations SHOULD error out.

> This extension MUST be marked as non-critical. Implementations MUST error out if the extension is not present AND the certificate is not self-signed.

Reviewer: I'm not sure 'error out' is good terminology here. Suggest 'Implementations must generate/create an error if the extension is not present...'

Author: ```suggestion
This extension MUST be marked as non-critical. Implementations MUST return an error if the extension is not present AND the certificate is not self-signed.
```

#### `subjectKeyIdentifier` Extension {#subject-key-id-ext}

The `subjectKeyIdentifier` extension identifies certificates that contain a particular public key. It can be used, for example, by control plane messages to identify which certificate to use for verification. The extension allows for overlapping control plane CA keys, for example during updates.

For the syntax and definition of the `subjectKeyIdentifier` extension, see {{RFC5280}}, section 4.2.1.2, and {{X.509}}, clause 9.2.2.2.

> This extension MUST be marked as non-critical. Implementations MUST error out if the extension is not present.

Reviewer: Same comment about 'error out'.

Author: ```suggestion
This extension MUST be marked as non-critical. Implementations MUST return an error if the extension is not present.
```

- `decipherOnly`: Not used.
- `keyCertSign`: The public key can be used to verify the CA signature on a control plane AS certificate.

**Important:** If a certificate’s public key is used to verify the signature of a control plane payload (`digitalSignature` attribute), it MUST be possible to trace back the private key used to sign the certificate. This is done by referencing the ISD-AS and the subject key identifier (via the `subjectKeyIdentifier` extension). For more information about the `subjectKeyIdentifier` extension, see [](#subject-key-id-ext).
Other attributes are not used.

If present, the `keyUsage` extension SHOULD be marked as "critical". That is, the `critical` Boolean attribute of this extension MUST be set to TRUE (the default is FALSE).
If a certificate’s public key is used to verify the signature of a control plane payload (`digitalSignature` attribute), it MUST be possible to trace back the private key used to sign the certificate. This is done by referencing the ISD-AS and the subject key identifier (via the `subjectKeyIdentifier` extension). For more information about the `subjectKeyIdentifier` extension (see [](#subject-key-id-ext)).

> **Note**: If a certificate extension is marked "critical", the public key in the certificate SHOULD only be used for the purpose set in the critical extension.

Author: > **Note**: If a certificate extension is marked "critical", the public key in the certificate SHOULD only be used for the purpose set in the critical extension.

@oncilla I removed this text above because I believe this is duplicating what RFC 5280 already states On KeyUsage:

> If the keyUsage extension is present, then the subject public key
MUST NOT be used to verify signatures on certificates or CRLs unless
the corresponding keyCertSign or cRLSign bit is set.

I think we could therefore piggyback on 5280 and completely remove this sentence on the purpose of the key.

Do you agree?

Author: He commented via email: SGTM

The attributes of the `keyUsage` extension define possible ways of using the public key. The attributes have the following meaning in SCION:

- `digitalSignature`: The public key can be used to verify the digital signature of a control plane payload.
- `keyCertSign`: The public key can be used to verify the CA signature on a control plane AS certificate.

Other attributes are not used.

If a certificate’s public key is used to verify the signature of a control plane payload (`digitalSignature` attribute), it MUST be possible to trace back the private key used to sign the certificate. This is done by referencing the ISD-AS and the subject key identifier (via the `subjectKeyIdentifier` extension). For more information about the `subjectKeyIdentifier` extension (see [](#subject-key-id-ext)).

> If present, the `keyUsage` extension's  `critical` attribute MUST be set to TRUE.

Author: @oncilla the draft's previous wording was very ambiguous about this (it included a MUST and a SHOULD). I think this is a MUST, correct? 

Or are there any cases where non-critical key usages are expected?

Reviewer 2: Quoting: https://www.rfc-editor.org/rfc/rfc5280.html#section-4.2.1.3

> Conforming CAs MUST include this extension in certificates that
   contain public keys that are used to validate digital signatures on
   other public key certificates or CRLs.  When present, conforming CAs
   SHOULD mark this extension as critical.
   
I think we tried to capture this and be in line with RFC 5280.

Author: Thank you, then this is a SHOULD. 

P.S. the conflict in the old text was: 
> SHOULD be marked as "critical"

Which somehow conflicts with: 
> `critical` Boolean attribute of this extension MUST be set to TRUE

The latest wording is: 
> When present, this extension SHOULD be marked as critical. In such case, the public key in the certificate SHOULD only be used for the purpose set in the critical extension.

The settings of the `basicConstraints` extension differ for each SCION Control Plane PKI certificate type. The next table shows the specifications per certificate type.

| Certificate Type                    | Root                   | CA                            | AS                            | Voting (regular and sensitive)    |
| ----------------------------------- | ---------------------- | ----------------------------- | ----------------------------- | --------------------------------- |
| *Attribute:*                        |                        |                               |                               |                                   |
| `basicConstraints` extension itself | MUST be present        | MUST be present               | SHOULD NOT be present         | SHOULD NOT be present             |
| `cA`                               | MUST be set to TRUE    | MUST be set to TRUE           | If the extension is present, this attribute MUST be set to FALSE | If the extension is present, this attribute MUST be set to FALSE |

> | `pathLenConstraint`                | SHOULD be set to "1", MUST be marked as "critical" | SHOULD be set to "0" (1), MUST be marked as "critical" | If the extension is present, this attribute MUST be absent. | If the extension is present, this attribute MUST be absent. |

Author: @oncilla is there any reason why this is a SHOULD rather than a MUST? Are there cases when the certificate path is longer than 1?

> `pathLenConstraint`                | SHOULD be set to "1", MUST be marked as "critical" | SHOULD be set to "0"

Reviewer 2: I think we tried to be in line with RFC 5280, but actually in the implementation we have this:

https://github.com/scionproto/scion/blob/68db9a02ff3f7bb0b3cc39217be001cfb2ac6d4f/pkg/scrypto/cppki/certs.go#L305

https://github.com/scionproto/scion/blob/68db9a02ff3f7bb0b3cc39217be001cfb2ac6d4f/pkg/scrypto/cppki/certs.go#L453-L455

So, actually we enforce it AFAICT.

Author: So shall I change it to MUST? 
"Must be set to 0" and "MUST be set to 1"

Author: I have to merge this to submit a new draft version and keep the ISE process going. I left a note about this SHOULD in https://github.com/scionassociation/scion-cppki_I-D/issues/132

Author: Dominik commented: 
MUST is good

Changing this in #135

~~~~

## TRC Updates {#update}

All non-base TRCs of an ISD are updates of the ISD's base TRC(s) and constitute a chain. Updates are categorized as regular or sensitive, depending on which payload fields are being modified.

> This section describes the rules that apply to updating a TRC in regard to the payload information contained in the TRC. Some rules are valid for both update types whilst some only apply to a regular or a sensitive TRC update. Based on the type of update, different sets of voters are needed to create a verifiable TRC update and the corresponding voting (signing) process is also described. Last, this section describes checks to verify a newly issued TRC.

Reviewer: I suggest 'Finally' before 'this section describes check.."

Author: ```suggestion
This section describes the rules that apply to updating a TRC in regard to the payload information contained in the TRC. Some rules are valid for both update types whilst some only apply to a regular or a sensitive TRC update. Based on the type of update, different sets of voters are needed to create a verifiable TRC update and the corresponding voting (signing) process is also described. Finally, this section describes checks to verify a newly issued TRC.
```

It is up to the ISD members to decide how the "casting a vote" procedure for updated TRCs will take place. Some ISDs make a distinction between regular and sensitive updates by dividing the regular and sensitive signing keys in different security classes, e.g. they keep the regular key in an online vault while the sensitive key would be stored offline. This way, the regular TRC update would lend itself to being automated (since the keys are accessible online) whereas the sensitive one would require manual actions to access the offline key. Other ISDs keep both regular and sensitive keys online and perform both updates automatically.

### TRC Update Verification {#trc-verification}

To verify a TRC update, the relying party MUST perform the following checks:

- Check that the specified update rules as described above are respected.

> - Check that all signatures are verifiable, and no superfluous signatures are attached.

Reviewer: I don't think a comma is needed here.

Author: ```suggestion
- Check that all signatures are verifiable and no superfluous signatures are attached.
```
